### PR TITLE
[SPARK-50382][CONNECT] Add documentation for general information on application development with/extending Spark Connect

### PR DESCRIPTION
**Pull request by vicennial**
### What changes were proposed in this pull request?

Adds a new page, `app-dev-spark-connect.md`, which is hyperlinked from the `Use Spark Connect in standalone applications` section in `spark-connect-overview`.

### Why are the changes needed?

There is a lack of documentation in the area of application development (with Spark Connect) especially so on extending Spark Connect with custom logic/libraries/plugins.

### Does this PR introduce _any_ user-facing change?

Yes, new page titled "Application Development with Spark Connect"

Render screenshot:
image

### How was this patch tested?

Local rendering

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

File: docs/app-dev-spark-connect.md [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+
+{% highlight python %}
+from pyspark.sql.connect.column import Expression
+from myxample.proto import ExamplePluginExpression
+import pyspark.sql.connect.proto as proto
+
+# Internal class that satisfies the interface by the Python client
+# of Spark Connect to generate the protobuf representation from
+# an instance of the expression.
+class ExampleExpression(Expression):
```
Comment (HyukjinKwon) [inline docs/app-dev-spark-connect.md:222]: nit but indentation would have to use 4 spaces instead of tabs

File: docs/app-dev-spark-connect.md [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+1. The Spark Connect protocol extension (blue box _Proto_ API)
+2. A Spark Connect Plugin…
+3. … with the application logic that extends Spark
+4. The client package that exposes the Spark Server Library application logic to the Spark Client
+Application, alongside PySpark or the Scala Spark Client. 
+<p style="text-align: center;">
+  <img src="img/extending-spark-connect-labelled.png" title="Figure 2: Labelled Architecture" alt="Extending Spark
+Connect Diagram - Labelled Steps" />
+</p> 
+#### (1) Spark Connect Protocol Extension
```
Comment (HyukjinKwon) [inline docs/app-dev-spark-connect.md:116]: Let's add a space above and below of those `####`. They aren't properly rendered in some markdown editors (as technically it's the standard to put spaces around them)

File: docs/app-dev-spark-connect.md [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+leveraged from everywhere. It can be embedded in modern data applications,
+in IDEs, Notebooks and programming languages.
+
+To learn more about Spark Connect, see [Spark Connect Overview](spark-connect-overview.html).
+
+# Redefining Spark Applications using Spark Connect
+
+With its decoupled client-server architecture, Spark Connect simplifies how Spark Applications are
+developed.
+The notion of Spark Client Applications and Spark Server Libraries are introduced as follows:
```
Comment (HyukjinKwon) [inline docs/app-dev-spark-connect.md:35]: Hmmm .. I feel a little bit awkward to call it "Libraries" .... I wonder if we should just remove ....

File: docs/app-dev-spark-connect.md [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+
+With Spark 3.4 and Spark Connect, the development of Spark Client Applications is simplified, and
+clear extension points and guidelines are provided on how to build Spark Server Libraries, making
+it easy for both types of applications to evolve alongside Spark. As illustrated in Fig.1, Spark
+Client applications connect to Spark using the Spark Connect API, which is essentially the
+DataFrame API and fully declarative.
+
+<p style="text-align: center;">
+  <img src="img/extending-spark-connect.png" title="Figure 1: Architecture" alt="Extending Spark
+Connect Diagram" />
```
Comment (HyukjinKwon) [inline docs/app-dev-spark-connect.md:51]:
```suggestion
Connect Diagram"/>
```

File: docs/app-dev-spark-connect.md [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+
+Spark Connect is available and supports PySpark and Scala
+applications. We will walk through how to run an Apache Spark server with Spark
+Connect and connect to it from a client application using the Spark Connect client
+library.
+
+A Spark Server Library consists of the following components, illustrated in Fig. 2:
+
+1. The Spark Connect protocol extension (blue box _Proto_ API)
+2. A Spark Connect Plugin…
```
Comment (HyukjinKwon) [inline docs/app-dev-spark-connect.md:108]: is `…` a typo? or incomplete? Below too.

File: docs/app-dev-spark-connect.md [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+3. … with the application logic that extends Spark
+4. The client package that exposes the Spark Server Library application logic to the Spark Client
+Application, alongside PySpark or the Scala Spark Client. 
+<p style="text-align: center;">
+  <img src="img/extending-spark-connect-labelled.png" title="Figure 2: Labelled Architecture" alt="Extending Spark
+Connect Diagram - Labelled Steps" />
```
Comment (HyukjinKwon) [inline docs/app-dev-spark-connect.md:114]:
```suggestion
Connect Diagram - Labelled Steps"/>
```

File: docs/app-dev-spark-connect.md [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+{% endhighlight %} 
+To make the example available in Python, the application developer provides a Python library that
+wraps the new expression and embeds it into PySpark. The easiest way to provide a function for any
+expression is to take a PySpark column instance as an argument and return a new Column instance
+with the expression applied. 
+
+{% highlight python %}
+from pyspark.sql.connect.column import Expression
+from myxample.proto import ExamplePluginExpression
+import pyspark.sql.connect.proto as proto
```
Comment (HyukjinKwon) [inline docs/app-dev-spark-connect.md:217]:
```suggestion
from pyspark.sql.connect.column import Expression
import pyspark.sql.connect.proto as proto

from myxample.proto import ExamplePluginExpression
```

Comment (HyukjinKwon) [reply docs/app-dev-spark-connect.md:217]: per PEP8

File: docs/app-dev-spark-connect.md (hunk 5 above)
Comment (vicennial) [reply docs/app-dev-spark-connect.md]: It was intentional to use it as a sort of continuation, but I see your point, it's a bit ambiguous. Updated text for clarity 👍

File: docs/app-dev-spark-connect.md (hunk 2 above)
Comment (vicennial) [reply docs/app-dev-spark-connect.md:117]: Thanks for the heads up!

File: docs/app-dev-spark-connect.md (hunk 1 above)
Comment (vicennial) [reply docs/app-dev-spark-connect.md:226]: Oops

**Comment by vicennial** [comment]
Thanks for the review @HyukjinKwon! I've addressed the feedback and updated the rendering in the PR description

**Review by grundprinzip — APPROVED** [review]
Thanks for doing the write up!

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by nchammas** [comment]
@grundprinzip - Does this work replace the WIP you had in #45340?

**Comment by vicennial** [comment]
@nchammas Not completely. There is more information in https://github.com/apache/spark/pull/45340 that makes quite a lot of sense to include, especially about the bits that explain the concepts of relations/expressions/commands. I've been meaning to integrate this information and haven't got into it yet but it's on my radar

**Pull request closed without merging**